src/H5system.c

```
@@ -1454,5 +1454,129 @@ HDqsort_context(void *base, size_t nel, size_t size, int (*compar)(const void *,
     /* Old BSD-style: context parameter comes before comparator function */
     qsort_r(base, nel, size, &wrapper, HDqsort_context_wrapper_func);
 #endif
+    return SUCCEED;
 }
 #endif
+
+/*
+ * HDqsort_fallback - Fallback qsort implementation for platforms without qsort_r/qsort_s
+ *
+ * For platforms that don't provide any reentrant qsort variant, this fallback uses
+ * thread-local storage (when thread-safety is
+ * enabled) or a global variable to store the comparator context, then uses standard qsort().
+ *
+ * The threadsafe version uses thread-local storage but does not support recursive sorting (a
+ * comparator calling HDqsort_r) as it would overwrite the previous context.
+ */
+#ifndef H5_HAVE_QSORT_REENTRANT
+
+typedef struct HDqsort_fallback_context_t {
+    int (*gnu_compar)(const void *, const void *, void *);
+    void *gnu_arg;
+} HDqsort_fallback_context_t;
+
+#ifdef H5_HAVE_THREADSAFE
+/* Thread-local storage approach for thread-safe builds
+ *
+ * Error Handling Note:
+ * HDqsort_fallback_key_init() still uses assertions because it's a callback with void signature.
+ * However, HDqsort_fallback() now returns herr_t and can propagate TLS operation failures
+ * through the normal HDF5 error stack.
+ */
+static H5TS_key_t  HDqsort_fallback_key;
+static H5TS_once_t HDqsort_fallback_key_once = H5TS_ONCE_INITIALIZER;
+
+static void
+HDqsort_fallback_key_init(void)
+{
+    herr_t ret = SUCCEED;
+
+    /* Create the thread-local storage key (no destructor needed) */
+    /* If this operation fails, it will be detected shortly during
+     * HDqsort_fallback when operations are attempted on the non-existent key */
+    H5TS_key_create(&HDqsort_fallback_key, NULL);
+
+    /* Assert that initialization succeeded - cannot propagate errors from here */
+    if (H5_UNLIKELY(ret < 0)) {
+        assert(false && "Failed to create TLS key for qsort fallback");
```

We _must_ avoid assertions in main library code for cases that can realistically occur, even if that means ignoring errors (though ideally we don't do that either).  In this case though, it looks like `ret` isn't assigned anyway, so this doesn't seem to be a useful check, unless `ret` should have been assigned from `H5TS_key_create`.

---

That said, I'd be much more in favor of simply dropping the thread-safe variant unless there's a good use case, and requiring `qsort_r` to exist for thread-safe builds, rather than bending over backwards a bit to add in error handling.

---

You're right, `ret` should have been assigned from the key creation. I've removed this, since like the comment says if key creation fails, that'll be detected immediately afterwards in `HDqsort_fallback()`.

---

> That said, I'd be much more in favor of simply dropping the thread-safe variant unless there's a good use case, and requiring `qsort_r` to exist for thread-safe builds, rather than bending over backwards a bit to add in error handling.

What exactly do you mean by requiring `qsort_r` to exist for threadsafe builds? I think that having the entire threadsafe build be impossible on systems that don't provide `qsort_r` directly seems too severe. As it stands, a threadsafe build should acquire the API lock on entry to the dataset routines that trigger `qsort_r` at tree create time anyway, so this should still be safe.

If the API lock isn't enough to handle the cases where the user wants a threadsafe build AND their system doesn't provide qsort_r, then I'd be more inclined to either fallback to not using the r-tree, or to acquire a special lock at tree creation time.

---

I think it's reasonable to require more modern features as the library progresses forward, but for the time being we could simply document the issues for now. For concurrency builds (as opposed to thread-safe builds), this may eventually become an issue.

---

In that case, the current commit should be ready, unless there's somewhere else that this should be documented.